File: 设计模式/设计模式/命令模式/单品/DanpinCommandManager.m
//
//  DanpinCommandManager.m
//  设计模式
//
//  Created by zachary spark on 2019/5/13.
//  Copyright © 2019 3g. All rights reserved.
//

#import "DanpinCommandManager.h"

@interface DanpinCommandManager()
@property(nonatomic,strong)NSMutableArray *commands;
@property(nonatomic,strong)DanPinSwitchCommand *OffCommand;
@property(nonatomic,strong)DanPinSetTimeCommand *SetTimeCommand;
@property(nonatomic,strong)DanPinSwitchModeCommand *SwitchModeCommand;

@end

@implementation DanpinCommandManager



-(instancetype)initSwitch:(DanPinSwitchCommand*)Off SetTime:(DanPinSetTimeCommand*)SetTime SwitchMode:(DanPinSwitchModeCommand*)SwitchMode{
    
    self = [super init];
    if(self){
        self.commands = [[NSMutableArray alloc]init];
        self.OffCommand = Off;
        self.SetTimeCommand = SetTime;
        self.SwitchModeCommand = SwitchMode;
    }
    return self;
}

-(void)toOff{
    [self.OffCommand execute];
    [self.commands addObject:self.OffCommand];
}

-(void)setTime:(NSTimeInterval)Interval{
    self.SetTimeCommand.interval = Interval;
    [self.SetTimeCommand execute];
    [self.commands addObject:self.SetTimeCommand];
}

-(void)switchMode:(DanPinDetectionMode)Mode{
    self.SwitchModeCommand.Mode = Mode;
    [self.SwitchModeCommand execute];
    [self.commands addObject:self.SwitchModeCommand];
}

//回退上一个命令
- (void)Undo{
    
    [[self.commands lastObject] execute];
    [self.commands removeLastObject];
    
}
//回退所有的命令
- (void)UndoAll{
    for (id<DanPinCommandProtocol> commad in self.commands) {
        [commad execute];
    }
    [self.commands removeAllObjects];
}


@end

File: 设计模式-Swift/设计模式-Swift/命令模式/基本实现/DanPinCommandManager.swift
//
//  DanPinCommandManager.swift
//  设计模式-Swift
//
//  Created by zachary spark on 2019/5/14.
//  Copyright © 2019 3g. All rights reserved.
//

//当命令太多的时候我们就需要了动态命令


import UIKit

class DanPinCommandManager: NSObject {

    let Off:DanPinSwitchCommand
    let SetTime:DanPinSetTimeCommand
    let SwitchMode:DanPinSwitchModeCommand
    
    var Commands:[DanPinCommandProtocol] = [DanPinCommandProtocol]()
    
    
    init(_ Off:DanPinSwitchCommand,_ SetTime:DanPinSetTimeCommand,_ SwitchMode:DanPinSwitchModeCommand) {
        
        self.Off = Off;
        self.SetTime = SetTime
        self.SwitchMode = SwitchMode
    }
    
    
    func toOff(){
        self.Off.execute()
        self.Commands.append(self.Off)
    }
    
    func setTime(_ Interval:TimeInterval){
        
        self.SetTime.Interval = Interval
        self.SetTime.execute()
        self.Commands.append(self.SetTime)
    }
    
    func switchMode(_ Mode:DanPinDetectionMode){
        
        self.SwitchMode.mode = Mode
        self.SwitchMode.execute()
        self.Commands.append(self.SetTime)
    }
    
    func Undo(){
        self.Commands.removeLast().execute()
    }
    
    func UndoAll(){
        
        for command in self.Commands {
            command.execute()
        }
        self.Commands.removeAll()
        
    }
    
}
